### ISSUE TITLE
Use of middlewares for `zod`

### ISSUE THREAD

**Issue by G0maa** [G0maa/my-school-server#32]
- As suggested by @Amr2812
- Reference: https://github.com/G0maa/my-school-server/blob/a8971d8341a0ff4440dc9c9abbc886de34d3aaae/src/controllers/student.ts#L23-L24
- and _all_ other controllers.
- Goal: Single middleware use, single schema to fulfill the above need.

**Comment by G0maa** [comment]
Revisiting.

I've been trying to adapt this to the current code, but I haven't been able to.


So far what I have got is
```ts
const validateSchemas =
  (schemas: AnyZodObject[]) =>
  (req: Request, _res: Response, next: NextFunction) => {
    // Error handling in errror-handler middleware
    // await/promise is needed only if you use **async** `refinements` or `transformers`
    for (const schema of schemas) {
      schema.parse({
        body: req.body,
        query: req.query,
        params: req.params,
      });
    }
    return next();
  };
```
And

```ts
export const ZStudentPost = z.object({
  body: ZStudent,
});
```

I cant use `transform` since it will invalidate the state of `req.body`:
```ts
export const ZUserPost = z
  .object({
    body: ZUser,
  })
  .transform((obj) => {
    return { ...obj, user: obj.body };
  });
  ```

Which means, either:
- Find a way to parse `req.body` twice populating it with the needed schema i.e. `req.body.user` & `req.body.student`
- Redesign schemas around routes themselves rather than entities.

**Comment by G0maa** [comment]
I used `ChatGPT` for 3-4 tries to refine the suggested code:

```ts
function validateAndExtract(schemas, properties) {
    return (req, res, next) => {
        try {
            let parsed = {};
            schemas.forEach(schema => {
                parsed = Object.assign(parsed, schema.parse(req.body));
            });
            properties.forEach(property => {
                req.body[property] = parsed[property];
            });
            next();
        } catch (err) {
            return res.status(400).json({ message: err.message });
        }
    }
}
```
i.e. to use
```ts
validateAndExtract([teacherSchema, userSchema], ['teacher', 'user'])

```

**Comment by G0maa** [comment]
- Revisited, tried & applied:
- Here it is used to verify entire request, but manually giving types to `req`, **this hugely depends on client sending different objects, user, student & studentDetails in req**
- https://github.com/G0maa/my-school-server/blob/576c72f15ca8e0440695c4eb724b1f794d07f6b9/src/controllers/teacher.ts#L55-L67
- Here it is too much to use a `middleware` for a small task:
https://github.com/G0maa/my-school-server/blob/576c72f15ca8e0440695c4eb724b1f794d07f6b9/src/controllers/teacher.ts#L92-L101
- And here you can't use it, `req.query` can't be sent as objects.
- https://github.com/G0maa/my-school-server/blob/576c72f15ca8e0440695c4eb724b1f794d07f6b9/src/controllers/teacher.ts#L28-L47

**Comment by G0maa** [comment]
- 3rd method
https://github.com/G0maa/my-school-server/blob/359143fb8524864d11340294aaa43a5e75f07a9f/src/controllers/fee.ts#L38-L56

**Comment by G0maa** [comment]
- For Reference, old:
https://github.com/G0maa/my-school-server/blob/def70826f7e5d43e8f683b0816e4e686d51e6192/src/validator/studyClass.validator.ts

**Comment by G0maa** [comment]
Solved with #45
